metricbeat/beater/reload.go

```diff
@@ -0,0 +1,252 @@
+package beater
+
+import (
+	"expvar"
+	"fmt"
+	"os"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/beats/libbeat/publisher"
+	"github.com/elastic/beats/metricbeat/mb"
+)
+
+var (
+	debugr        = logp.MakeDebug("reloader")
+	configReloads = expvar.NewInt("metricbeat.config.reloads")
+	moduleStarts  = expvar.NewInt("metricbeat.config.module.starts")
+	moduleStops   = expvar.NewInt("metricbeat.config.module.stops")
+	moduleRunning = expvar.NewInt("metricbeat.config.module.running")
+
+	ErrNoNewFiles = errors.New("no new files found")
+)
+
+type ConfigReloader struct {
+	registry RunningRegistry
+	config   ReloaderConfig
+	client   func() publisher.Client
+	done     chan struct{}
+	wg       sync.WaitGroup
+}
+
+type RunningRegistry struct {
+	List  map[uint64]ModuleRunner
+	mutex sync.Mutex
+}
+
+func NewConfigReloader(config ReloaderConfig, p publisher.Publisher) ConfigReloader {
+
+	// TODO: Is there any downside to call connect for each new module?
+	client := func() publisher.Client { return p.Connect() }
+
+	return ConfigReloader{
+		registry: NewRunningRegistry(),
+		config:   config,
+		client:   client,
+		done:     make(chan struct{}),
+	}
+}
+
+func (r *ConfigReloader) Start() {
+
+	r.wg.Add(1)
+	defer r.wg.Done()
+
+	lastScan := time.Time{}
+
+	for {
+		select {
+		case <-r.done:
+			logp.Info("Monitoring stopped")
+			return
+		case <-time.After(r.config.Period):
+
+			debugr("Scan for new config files")
+			configReloads.Add(1)
+
+			lastScanTemp := time.Now()
+			newFiles, err := FetchFiles(r.config.Path, lastScan)
+			if err != nil {
+				if err != ErrNoNewFiles {
+					logp.Err("Error fetching files: %v", err)
+				}
+				// Continue on error or no new files
+				continue
+			}
+
+			lastScan = lastScanTemp
```

Comment:
Even thought that was lost during refactoring and initially the purpose was to only update lastScan if no error, I think it is still correct like this. If there are no file updates, it does not matter if the comparison of the files is to the most recent lastScan or one of the previous where no files were found. The resulting outcome should be the same and it keeps the code simpler. I will add a comment to the code.